### PR TITLE
DBM-APM: Selectively pre/append sql comments depending on DBMS typeq

### PR DESCRIPTION
**Pull request by jmeunier28**
# What Does This Do

This PR updates the logic for the APM-DBM link feature introduced in https://github.com/DataDog/dd-trace-java/pull/4847. It changes from prepending the sql comments, containing the APM trace context, to appending them to the raw sql for **SQL Server driver implementations only**. This requires no changes on the DBM side of things in either the agent or our backend.

Jira ticket: [DBMON-2759](https://datadoghq.atlassian.net/browse/DBMON-2759)

# Motivation

This fixes an issue, which has been reported by several customers that executing `CallableStatement`s with the following syntax causes exceptions to be thrown when enabling APM-DBM link:

> The customer’s code base makes a CallableStatement statementvariable = connection.prepareCall("{call some_function_namehere(?,?,?) }";) and whenever both of these are active, the Java application starts throwing syntax error at or near "{" over and over.

We were able to reproduce this, using the latest version of the java tracer && an internal application. It seems to be caused by the SQL comments we inject being _prepended_ instead of _appended_.  For example, the issue of prepending the comment was reproduced with a simple java application (without APM-DBM link enabled), which did the following:

```java
  final String storedProcedureCall = "/*test commnet*/ {call foo(?, ?)}";
  CallableStatement callableStmt = con.prepareCall(storedProcedureCall);
  callableStmt.setInt(1,10);
  callableStmt.setInt(2,11);
  callableStmt.execute();
  callableStmt.close(); 
```

... this throws the same exception with message `Incorrect syntax near '{'.`

# Proof of Fix

After doing a custom build of this tracer code, we deployed our test app, which runs a combination of `PreparedStatement`/`Statements`/`CallableStatement` & no longer saw the `Incorrect syntax near '{'.` exception being thrown. 

We were also able to confirm that injection is still working properly, e.g: our events are being properly decorated:

Screen Shot 2023-10-12 at 5 26 06 PM

Screen Shot 2023-10-12 at 5 26 15 PM

# Expected DBM Behavior for Stored Procedures

Stored Procedures aren't supported for the APM-DBM usecase (unless the same app that is calling them is the thing that created the stored procedure & the linking was enabled). Otherwise, we wouldn't see the comments if someone had code that was simply calling `EXEC proc` or `{call proc}`, since the actual text we see in the activity code is either:

1. The entire stored proc text e.g.:
```tsql
CREATE PROCEDURE foo
BEGIN
-- do stuff
END;
```
2. The [currently executing query](https://github.com/DataDog/integrations-core/blob/master/sqlserver/datadog_checks/sqlserver/activity.py#L55-L60) in the context of the stored proc

Neither, contains the text `EXEC proc`, which is what would have the comment injected via the tracer.

Obviously, this is of secondary concern here. Updating the tracer to append the comment will fix this edge case being reported by customers. **We really don't want the feature to cause users to see exceptions being thrown in their production code.**

[DBMON-2759]: https://datadoghq.atlassian.net/browse/DBMON-2759?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java [hunk 1]
```diff
@@ -73,6 +73,8 @@ public DBMCompatibleConnectionInstrumentation() {
     "com.informix.jdbc.IfmxConnection",
   };
 
+  private static final String locationMode = "prepend";
```
Comment (dougqh) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java:76]: It probably makes more sense for this to be a boolean or an enum.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java [hunk 2]
```diff
@@ -19,23 +19,28 @@ public class SQLCommenter {
   private static final char EQUALS = '=';
   private static final char COMMA = ',';
   private static final char QUOTE = '\'';
+  private static final String OPEN_COMMENT_APPEND = " /*";
```
Comment (dougqh) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java:22]: Why are there two constants for OPEN_COMMENT & OPEN_COMMENT_APPEND?
I also see that there's a use of the string literal as "/*" in the code as well.

I think having 3 of same thing is making the code a bit harder to follow.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java [hunk 3]
```diff
@@ -19,23 +19,28 @@ public class SQLCommenter {
...
   private static final String OPEN_COMMENT = "/*";
   private static final String CLOSE_COMMENT = "*/ ";
+  private static final String CLOSE_COMMENT_APPEND = "*/";
   private static final int INITIAL_CAPACITY = computeInitialCapacity();
+  private static final String PREPEND = "prepend";
```
Comment (dougqh) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java:27]: I'd replaced these with an enum or a boolean for the two different modes.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java [hunk 4]
```diff
@@ -44,6 +49,16 @@ public static String inject(
     final String version = config.getVersion();
     final int commentSize = capacity(traceParent, parentService, dbService, env, version);
     StringBuilder sb = new StringBuilder(sql.length() + commentSize);
+    if (locationMode.equals(APPEND)) {
+      sb.append(sql);
+      sb.append(OPEN_COMMENT_APPEND);
+      toComment(sb, injectTrace, parentService, dbService, env, version, traceParent);
+      if (sb.length() == OPEN_COMMENT_APPEND.length() + sql.length()) {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java:56]: It isn't immediately obvious to me what this check is doing.
I'm guessing that it is checking whether the comment was able to be injected, but it is hard to tell.

If that's the case, I'd prefer to see toComment return a boolean indicating success.
That would make the code flow easier to follow.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java [hunk 5]
```diff
@@ -44,6 +49,16 @@ public static String inject(
     final String version = config.getVersion();
     final int commentSize = capacity(traceParent, parentService, dbService, env, version);
     StringBuilder sb = new StringBuilder(sql.length() + commentSize);
+    if (locationMode.equals(APPEND)) {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java:52]: I think I'd prefer to see this as an if / else - rather than an if with an early return.
To me, that would make the code easier to understand.

File: dd-java-agent/instrumentation/jdbc/src/test/groovy/SQLCommenterTest.groovy [hunk 6]
```diff
@@ -23,26 +23,47 @@ class SQLCommenterTest extends AgentTestRunner {
...
-    "/*customer-comment*/ SELECT * FROM foo"                                                      | "SqlCommenter" | "Test" | "my-service" | "TestVersion" | false       | null                                                      | "/*ddps='SqlCommenter',dddbs='my-service',dde='Test',ddpv='TestVersion'*/ /*customer-comment*/ SELECT * FROM foo"
-    "/*traceparent"                                                                               | "SqlCommenter" | "Test" | "my-service" | "TestVersion" | false       | null                                                      | "/*ddps='SqlCommenter',dddbs='my-service',dde='Test',ddpv='TestVersion'*/ /*traceparent"
+    query                                                                                         | ddService      | ddEnv  | dbService    | ddVersion     | injectTrace | locationMode                             | traceParent                                               | expected
```
Comment (dougqh) [inline dd-java-agent/instrumentation/jdbc/src/test/groovy/SQLCommenterTest.groovy:26]: The diff is hard to read.  Is there a test here that tests the edge case where the injection fails?

**Review by dougqh — CHANGES_REQUESTED** [review]
I find the logic hard to follow in places.
I'd like to see a bit of clean-up.

I also think the append vs prepend could be better handled with a boolean or an enum.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java (hunk 4 above)
Comment (jmeunier28) [reply dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java:56]: yeah sure, that makes sense

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java (hunk 1 above)
Comment (mcculls) [reply dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java:76]: boolean would be preferred in this case - avoids having to list all the enum classes as helpers

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java [hunk 7]
```diff
@@ -124,7 +127,7 @@ public static String onEnter(
         final DBInfo dbInfo =
             JDBCDecorator.parseDBInfo(
                 connection, InstrumentationContext.get(Connection.class, DBInfo.class));
-        sql = SQLCommenter.inject(sql, DECORATE.getDbService(dbInfo));
+        sql = SQLCommenter.inject(sql, DECORATE.getDbService(dbInfo), appendComment);
```
Comment (mcculls) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java:130]: Have you considered providing two methods `SQLCommenter.append` and `SQLCommenter.prepend` which each just call `SQLCommenter.inject` with the appropriate boolean flag?

I think that would be much more self-documenting.

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java [hunk 8]
```diff
@@ -19,23 +19,29 @@ public class SQLCommenter {
...
   private static final char QUOTE = '\'';
+  private static final char SPACE = ' ';
   private static final String OPEN_COMMENT = "/*";
-  private static final String CLOSE_COMMENT = "*/ ";
+  private static final String CLOSE_COMMENT = "*/";
   private static final int INITIAL_CAPACITY = computeInitialCapacity();
 
-  public static String inject(final String sql, final String dbService) {
-    return inject(sql, dbService, null, false);
+  public static String append(final String sql, final String dbService) {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/SQLCommenter.java:27]: I like the splitting of the functions.

**Review by dougqh — APPROVED** [review]

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java (hunk 7 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DBMCompatibleConnectionInstrumentation.java]: Yes, that was a nice improvement.  Thanks for doing that.

**Review by am312 — APPROVED** [review]

**Comment by jmeunier28** [comment]
the test failures are unrelated to my PR && is fixed in here https://github.com/DataDog/dd-trace-java/pull/6177

Going to merge this since we have confidence based on the relevant unit / system tests passing
